pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,234 @@
+			fromTime = lastExecTime
+		} else {
+			fromTime = time.Now().Add(-jitter)
+		}
+	}
+
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s, catchupTime, err)
+		} else {
+			w.snapshot.UpdateLastExecutionTime(nameOfSchedule, catchupTime)
+		}
+	}
+
+	return nil
+}
+
+func fire(ctx context.Context, executionManager mgInterfaces.ExecutionInterface, scheduledTime time.Time,
+	s models.SchedulableEntity) error {
+
+	literalsInputMap := map[string]*core.Literal{}
+	literalsInputMap[s.KickoffTimeInputArg] = &core.Literal{
+		Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: &core.Primitive{
+						Value: &core.Primitive_Datetime{
+							Datetime: timestamppb.New(scheduledTime),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	executionRequest := admin.ExecutionCreateRequest{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19],
+		Spec: &admin.ExecutionSpec{
+			LaunchPlan: &core.Identifier{
+				ResourceType: core.ResourceType_LAUNCH_PLAN,
+				Project:      s.Project,
+				Domain:       s.Domain,
+				Name:         s.Name,
+				Version:      s.Version,
+			},
+			Metadata: &admin.ExecutionMetadata{
+				Mode:        admin.ExecutionMetadata_SCHEDULED,
+				ScheduledAt: timestamppb.New(scheduledTime),
+			},
+			// No dynamic notifications are configured either.
+		},
+		// No additional inputs beyond the to-be-filled-out kickoff time arg are specified.
+		Inputs: &core.LiteralMap{
+			Literals: literalsInputMap,
+		},
+	}
+
+	_, err := executionManager.CreateExecution(context.Background(), executionRequest, scheduledTime)
+	if err != nil {
+		logger.Error(ctx, "failed to create  execution create  request due to %v", err)
+		return err
+	}
+	return nil
+}
+
+func (w *workflowExecutor) Run() {
+	ctx := context.Background()
+	c, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		panic(fmt.Errorf("unable to run the workflow executor after reading the schedules due to %v", err))
+	}
+	// Start the go routine to write the snapshot map to the configured snapshot file
+	go w.CheckPointState(ctx)
+	catchUpTill := time.Now()
+	err = w.CatchUpAllSchedules(ctx, c.Entities, catchUpTill)
+	if err != nil {
+		panic(fmt.Errorf("unable to catch up on the schedules due to %v", err))
+	}
+
+	defer logger.Infof(ctx, "Exiting Workflow executor")
+	for true {
+		for _, s := range c.Entities {
+			funcRef := func() {
+				nameOfSchedule := w.goGfInterface.GetScheduleName(s)
+				lastT := w.snapshot.GetLastExecutionTime(nameOfSchedule)
+				n := time.Now()
+				err = w.CatchUpSingleSchedule(ctx, s, n)
```

Comment:
I'm confused, doesn't the call above to w.CatchUpAllSchedules (line 171) call CatchUpSingleSchedule for each schedule entity?

Reply:
So the captupallschedules happens in the main scheduler routine during bootup
and uses the time.Now when this is executed.
```
catchUpTill := time.Now()
err = w.CatchUpAllSchedules(ctx, c.Entities, catchUpTill)
``
So here all the schedules are caught up to catchUpTill


But when this function is invoked because of cron / regular cadence
```
funcRef := func() {
				nameOfSchedule := w.goGfInterface.GetScheduleName(s)
				lastT := w.snapshot.GetLastExecutionTime(nameOfSchedule)
				n := time.Now()
				err = w.CatchUpSingleSchedule(ctx, s, n)
				if err != nil {
					logger.Errorf(ctx, "unable to catch on schedule %v from lastT %v to %v with jitter configured due to %v", s.Name, lastT, n, err)
				}
			}
```

At that time `n := time.Now()` would be different and we will catch the schedule from `lastT to n`

Does that answer your question?

Reply:
i think so, thanks